service/raft/group0_state_machine.cc

```diff
@@ -317,12 +319,24 @@ future<> group0_state_machine::transfer_snapshot(raft::server_id from_id, raft::
             &_mm._messaging, addr, as, from_id, service::raft_snapshot_pull_params{std::move(tables)});
 
         tables = std::vector<table_id>();
-        tables.reserve(auth_tables.size());
 
-        for (const auto& schema : auth_tables) {
-            tables.push_back(schema->id());
+        // for the normal (non-topology) raft snapshot, we send an empty vector parameter
+        // and let the sender decide which tables to send.
+        // it is possible the other node doesn't support the empty parameter mode, or
+        // maybe it supports it but we don't know it yet, so in this case we use the
+        // legacy rpc, setting the parameter with tables we want, and the sender may
+        // add additional tables if needed.
+
+        if (!_snapshot_rpc_receiver_side_enabled) {
```

Comment:
Why didn't you move `topology`, `topology_requests` and `cdc_generattion_v3` under the same mechanism?

Also @piodul do you remember when do enable features when joining a cluster? Will this feature be enabled if we join updated cluster and disable otherwise?

Reply:
After receiving an accept response from the topology coordinator, the joining node performs a group0 read barrier. This causes the group0 state to be reloaded, and this is when features are marked as enabled on the node for the first time.

Specifically, if the joining node requests snapshot transfer as a part of the read barrier, it will send the RPC in the legacy way.

Reply:
I kept the topology snapshot separate because it's handled differently, so I thought it's simpler and makes sense, and also helps to keep the rpc size not too big.

Reply:
> After receiving an accept response from the topology coordinator, the joining node performs a group0 read barrier. This causes the group0 state to be reloaded, and this is when features are marked as enabled on the node for the first time.
> 
> Specifically, if the joining node requests snapshot transfer as a part of the read barrier, it will send the RPC in the legacy way.

But then it will miss some data if a feature is already enabled, no?

Reply:
So it looks like the code should work this way: An upgraded node should be ready to accept new snapshot (containing all the table) no matter which kind of RPC it sent. When upgraded node receives a snapshot request it checks upgrade status and if upgraded it always sends everything, otherwise it sends only requested tables). @piodul makes sense?

Reply:
Makes sense to me.

Reply:
but RPC size limitation is important to consider here, is there any limit? because if we ever hit it we'll need another refactor as we can't "slice" snapshot with new RPC as we do now.

btw I agree that `topology, topology_requests and cdc_generattion_v3` should be unififed too

Reply:
RPC has no practical size limitation, but huge RPC has other issues. If the snapshot will become really large we wil l need to move to streaming to transfer it.

Reply:
I changed it to one rpc with all the tables.